Components/ANSIescapeCodeHandler.cpp

```diff
@@ -0,0 +1,248 @@
+/****************************************************************************
+**
+** Copyright (C) 2015 Petar Perisin <petar.perisin@gmail.com>
+** Contact: http://www.qt.io/licensing
+**
+** This file is part of Qt Creator.
+**
+** Commercial License Usage
+** Licensees holding valid commercial Qt licenses may use this file in
+** accordance with the commercial license agreement provided with the
+** Software or, alternatively, in accordance with the terms contained in
+** a written agreement between you and The Qt Company.  For licensing terms and
+** conditions see http://www.qt.io/terms-conditions.  For further information
+** use the contact form at http://www.qt.io/contact-us.
+**
+** GNU Lesser General Public License Usage
+** Alternatively, this file may be used under the terms of the GNU Lesser
+** General Public License version 2.1 or version 3 as published by the Free
+** Software Foundation and appearing in the file LICENSE.LGPLv21 and
+** LICENSE.LGPLv3 included in the packaging of this file.  Please review the
+** following information to ensure the GNU Lesser General Public License
+** requirements will be met: https://www.gnu.org/licenses/lgpl.html and
+** http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
+**
+** In addition, as a special exception, The Qt Company gives you certain additional
+** rights.  These rights are described in The Qt Company LGPL Exception
+** version 1.1, included in the file LGPL_EXCEPTION.txt in this package.
+**
+****************************************************************************/
+
+#include "ANSIescapeCodeHandler.h"
+
+namespace Utils {
+
+/*!
+    \class Utils::AnsiEscapeCodeHandler
+    \brief The AnsiEscapeCodeHandler class parses text and extracts ANSI escape codes from it.
+    In order to preserve color information across text segments, an instance of this class
+    must be stored for the lifetime of a stream.
+    Also, one instance of this class should not handle multiple streams (at least not
+    at the same time).
+    Its main function is parseText(), which accepts text and default QTextCharFormat.
+    This function is designed to parse text and split colored text to smaller strings,
+    with their appropriate formatting information set inside QTextCharFormat.
+    Usage:
+    \list
+    \li Create new instance of AnsiEscapeCodeHandler for a stream.
+    \li To add new text, call parseText() with the text and a default QTextCharFormat.
+        The result of this function is a list of strings with formats set in appropriate
+        QTextCharFormat.
+    \endlist
+*/
+
+AnsiEscapeCodeHandler::AnsiEscapeCodeHandler() :
+    m_previousFormatClosed(true)
+{
+}
+
+static QColor ansiColor(uint code)
+{
+    if (code < 8) return QColor();
```

Comment:
Use `R_EXPECT` here, albeit this tightly couples this component with the rest of our code base making it harder to pull updates to this source from Qt again.
https://github.com/enigma-dev/RadialGM/blob/b3cc0ac81a2011c300878b1f4893e92727e729f4/Components/Logger.h#L29